KEP-4671, KEP-5598: Introducing Workload Aware Scheduling blog post (#53012)
kubernetes/website: content/en/blog/_posts/2025-xx-xx-gang-scheduling.md

---
layout: blog
draft: true
title: "Kubernetes v1.35: Introducing Workload Aware Scheduling"
date: 2025-XX-XX
slug: introducing-workload-aware-scheduling
author: >
  Maciej Skoczeń (Google)
  Dominik Marciński (Google)
---

Scheduling large workloads is a much more complex and fragile operation than scheduling a single Pod,
as it requires repeating the same process dozens of times for each Pod independently,

> even when the Pods that are part of such a workload are identical.

Reviewer: ```suggestion
At the same time the Pods that are part of such a workload are very often identical, which completely changes how the scheduling process should look like.
```

There are many custom schedulers adapted to perform workload scheduling efficiently,
but considering how common and important workload scheduling is to Kubernetes users,
it's high time to make workloads a first-class citizen for `kube-scheduler` and support them natively.

## Workload aware scheduling

We are happy to announce the kickoff of the Workload Aware Scheduling effort that is aiming
to improve scheduling and management of workloads. The effort will span over many SIGs and releases,

> and is supposed to gradually expand capabilities of the system toward reaching the North Star goal.

Reviewer: ```suggestion
and is supposed to gradually expand capabilities of the system toward reaching the North Star goal, which is seamless workload scheduling and management in Kubernetes including, but not limited to, preemption and autoscaling.
```

In Kubernetes v1.35 we introduce the **Workload API** that will be used to describe the desired shape
as well as scheduling and runtime behavior of the workload. It comes with an initial implementation

> of **Gang Scheduling** that instructs the kube-scheduler to schedule gang Pods in the "all-or-nothing" fashion.

Reviewer 2: ```suggestion
of *Gang Scheduling* that instructs the `kube-scheduler` to schedule gang Pods in the "all-or-nothing" fashion.
```

Finally, we improved scheduling of identical Pods (that typically make a gang) to speed up the process
thanks to the *Opportunistic Batching* feature.

> ## Workload API

Reviewer 3: Thanks!

The new Workload API resource is part of the `scheduling.k8s.io/v1alpha1`
{{< glossary_tooltip text="API group" term_id="api-group" >}}.
This resource acts as a structured manifest for the scheduling requirements of a multi-Pod application.
While user-facing workloads like Jobs define what to run, the Workload resource
determines how a group of Pods should be scheduled.

A Workload allows you to define a group of Pods and apply a scheduling policy to them.
Here is what a Gang Scheduling configuration looks like. You can define a `podGroup` named `workers`
and apply the `gang` policy with a `minCount` of 4.

```yaml
apiVersion: scheduling.k8s.io/v1alpha1
kind: Workload
metadata:
  name: training-job-workload
  namespace: some-ns
spec:
  podGroups:
  - name: workers
    policy:
      gang:
        # The gang is schedulable only if 4 pods can run at once
        minCount: 4
```

When you create your Pods, you simply link them to this Workload using the new `workloadRef` field:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: worker-0
  namespace: some-ns
spec:
  workloadRef:
    name: training-job-workload
    podGroup: workers
  ...
```

> ## How gang scheduling works

Author: Added a paragraph emphasizing that

> ## How Gang Scheduling Works

Reviewer 2: Here would be useful to have an example of the behavior. 
If you configure this way, and these conditions are met then this will happen...

> ## How Gang Scheduling Works

Author: Do you mean modifying the current gang scheduling steps or adding a standalone example after the high level description?

   instead of binding them to nodes immediately, the Pods wait at a `Permit` gate.

3. The scheduler checks if it has found valid nodes for the entire gang (at least the `minCount`).
   * If there is room for the gang, the gate opens, and all Pods are bound to nodes.
   * If only a subset of the gang pods was successfully scheduled within a timeout (set to 5 minutes),
     the scheduler rejects *all* of them. They go back to the queue,
     freeing up resources for other jobs that *can* run.

## Opportunistic Batching

In addition to explicit gang scheduling, v1.35 introduces **Opportunistic Batching**.
This is a Beta feature that improves scheduling latency for identical Pods.

Unlike Gang Scheduling, this feature does not require the Workload API
or any explicit opt-in on the user's part. It works opportunistically within the scheduler
by identifying Pods that have identical scheduling requirements (container images, resource requests,
affinities, etc.). When the scheduler processes a Pod, it can reuse the feasibility calculations
for subsequent identical Pods in the queue, significantly speeding up the process.

### Restrictions

> Opportunistic Batching works under specific conditions. All fields used by the kube-scheduler

Reviewer 2: ```suggestion
Opportunistic Batching works under specific conditions. All fields used by the `kube-scheduler`
```

to find a placement must be identical between Pods. Additionally, using
[topology spread constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/),
[DRA ResourceClaims](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/)
or having [DRAExtendedResource](/docs/reference/command-line-tools-reference/feature-gates/DRAExtendedResource)
feature gate enabled, disables the batching mechanism for those Pods to ensure correctness.

Note that you may need to review your `kube-scheduler` configuration
to ensure default topology spread constraints are not implicitly disabling batching for your workloads.

See the [docs](TODO) for more details about restrictions.

## The north star vision

The scope of Workload Aware Scheduling is broad. These features are just the first steps.
In the near future, the effort aims to tackle:

* Single-cycle Gang Scheduling

> * Topology Aware Scheduling

Reviewer: ```suggestion
* Multi-node DRA and Topology Aware Scheduling
```

to improve scheduling and management of workloads. The effort will span over many SIGs and releases,
and is supposed to gradually expand capabilities of the system toward reaching the North Star goal,
which is seamless workload scheduling and management in Kubernetes including,
but not limited to, preemption and autoscaling.

Starting with Kubernetes v1.35, we introduce the Workload API that will be used to describe the desired shape
as well as scheduling and runtime behavior of the workload. It comes with an initial implementation
of *Gang Scheduling* that instructs the `kube-scheduler` to schedule gang Pods in the "all-or-nothing" fashion.
Finally, we improved scheduling of identical Pods (that typically make a gang) to speed up the process
thanks to the *Opportunistic Batching* feature.

## Workload API

The new Workload API resource is part of the `scheduling.k8s.io/v1alpha1`
{{< glossary_tooltip text="API group" term_id="api-group" >}}.
This resource acts as a structured manifest for the scheduling requirements of a multi-Pod application.
While user-facing workloads like Jobs define what to run, the Workload resource

> determines how a group of Pods should be scheduled and managed.

Reviewer 4: "and managed" can be misleading.
The lifecycle of the workload will still be managed by corresponding controllers.

What scheduling will be doing is "managing the placement of the workload for its whole lifecycle".

Author: Rephrased

---
layout: blog
draft: true
title: "Kubernetes v1.35: Introducing Workload Aware Scheduling"
date: 2025-XX-XX
slug: introducing-workload-aware-scheduling
author: >
  Maciej Skoczeń (Google)
  Dominik Marciński (Google)
---

Scheduling large workloads is a much more complex and fragile operation than scheduling a single Pod,
as it often requires considering all Pods together instead of scheduling each one independently.
At the same time, the Pods that are part of such a workload are very often identical
from the scheduling perspective, which fundamentally changes how this process should look.

There are many custom schedulers adapted to perform workload scheduling efficiently,
but considering how common and important workload scheduling is to Kubernetes users,
it's high time to make workloads a first-class citizen for `kube-scheduler` and support them natively.

## Workload Aware Scheduling

> We are happy to announce the kickoff of the Workload Aware Scheduling effort that is aiming

Reviewer 3: ```suggestion
We are happy to announce the kickoff of the _orkload aware scheduling_ effort that is aiming
```

ℹ️ This sounds implausible. Usually you don't kick things off after the release; this will be published a few days **later** than the v1.35 release has shipped.

to improve scheduling and management of workloads. The effort will span over many SIGs and releases,
and is supposed to gradually expand capabilities of the system toward reaching the North Star goal,
which is seamless workload scheduling and management in Kubernetes including,
but not limited to, preemption and autoscaling.

Starting with Kubernetes v1.35, we introduce the Workload API that will be used to describe the desired shape
as well as scheduling-oriented requirements of the workload. It comes with an initial implementation
of *Gang Scheduling* that instructs the `kube-scheduler` to schedule gang Pods in the "all-or-nothing" fashion.
Finally, we improved scheduling of identical Pods (that typically make a gang) to speed up the process

> thanks to the *Opportunistic Batching* feature.

Reviewer 5: ```suggestion
thanks to the *opportunistic batching* feature.
```

## Workload API

The new Workload API resource is part of the `scheduling.k8s.io/v1alpha1`
{{< glossary_tooltip text="API group" term_id="api-group" >}}.

> This resource acts as a structured manifest for the scheduling requirements of a multi-Pod application.

Reviewer 3: ```suggestion
This resource acts as a structured, machine-readable definition of the scheduling requirements of a multi-Pod application.
```

While user-facing workloads like Jobs define what to run, the Workload resource
determines how a group of Pods should be scheduled and how its placement should be managed
throughout its lifecycle.

A Workload allows you to define a group of Pods and apply a scheduling policy to them.
Here is what a Gang Scheduling configuration looks like. You can define a `podGroup` named `workers`
and apply the `gang` policy with a `minCount` of 4.

```yaml
apiVersion: scheduling.k8s.io/v1alpha1
kind: Workload
metadata:
  name: training-job-workload
  namespace: some-ns
spec:
  podGroups:
  - name: workers
    policy:
      gang:
        # The gang is schedulable only if 4 pods can run at once
        minCount: 4
```

> When you create your Pods, you simply link them to this Workload using the new `workloadRef` field:

Reviewer 3: ```suggestion
When you create your Pods, you link them to this Workload using the new `workloadRef` field:
```

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: worker-0
  namespace: some-ns
spec:
  workloadRef:
    name: training-job-workload
    podGroup: workers
  ...
```

## How gang scheduling works

The `gang` policy enforces "all-or-nothing" placement. Without gang scheduling,
a job might be partially scheduled, consuming resources without being able to run,
leading to resource wastage and potential deadlocks.

When you create Pods that are part of a gang-scheduled `PodGroup`, the scheduler's `GangScheduling`
plugin manages the lifecycle independently for each `PodGroup` (or replica key):

1. When you create your Pods, the scheduler blocks them from scheduling,

>    until the number of pending Pods meets your `minCount`.

Reviewer 3: ```suggestion
   until the number of pending Pods in that group meets your `minCount`.
```

2. Once enough Pods arrive, the scheduler tries to place them. However,
   instead of binding them to nodes immediately, the Pods wait at a `Permit` gate.

3. The scheduler checks if it has found valid assignments for the entire gang (at least the `minCount`).

>    * If there is room for the gang, the gate opens, and all Pods are bound to nodes.

Reviewer 3: Is this a gang or a group?

   * If only a subset of the gang pods was successfully scheduled within a timeout (set to 5 minutes),
     the scheduler rejects *all* of them. They go back to the queue,

>      freeing up resources for other jobs that *can* run.

Reviewer 3: ```suggestion
     the scheduler rejects **all** of the Pods in the gang.
     They go back to the queue, freeing up the reserved resources for other workloads.
```

We shouldn't assume that the rival workloads are Jobs.

Note that while this is a first implementation, the gang scheduling algorithm

> will be improved and expanded in future releases, including a single-cycle scheduling

Reviewer 3: ```suggestion
We'd like to point out that that while this is a first implementation, the Kubernetes project firmly
intends to improve and expand the gang scheduling algorithm in future releases.
Benefits we hope to deliver include a single-cycle scheduling
```

If making a _promise_ about an upcoming feature that looks like a commitment (eg: it's in an official project blog post), you should either:

- reword it (easy option)
- go to SIG Architecture, or even Steering, and get them to commit to that improvement happening, to the extend that we would refocus efforts to meet the commitment

The easy option is: don't make the commitment. Write it as an ambition instead.

Reviewer: @wojtek-t I'm not sure what are the best practices for the official blogs, but we indeed could skip listing the "north star vision" section and just say "stay tuned" linking to https://github.com/kubernetes/kubernetes/issues/132192.

The problem with the list is that it may be not super clear what exactly these terms mean, so it's even not a matter of commitment, but also precision what we "promise"

phase for a whole gang, workload-level preemption, and more, moving towards the North Star goal.

## Opportunistic batching

In addition to explicit gang scheduling, v1.35 introduces *Opportunistic Batching*.
This is a Beta feature that improves scheduling latency for identical Pods.

> Unlike Gang Scheduling, this feature does not require the Workload API

Reviewer 5: ```suggestion
Unlike gang scheduling, this feature does not require the Workload API
```

or any explicit opt-in on the user's part. It works opportunistically within the scheduler
by identifying Pods that have identical scheduling requirements (container images, resource requests,
affinities, etc.). When the scheduler processes a Pod, it can reuse the feasibility calculations
for subsequent identical Pods in the queue, significantly speeding up the process.

### Restrictions

> Opportunistic Batching works under specific conditions. All fields used by the `kube-scheduler`

Reviewer 5: ```suggestion
Opportunistic batching works under specific conditions. All fields used by the `kube-scheduler`
```

to find a placement must be identical between Pods. Additionally, using
[topology spread constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/),
[DRA ResourceClaims](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/)

> or having [DRAExtendedResource](/docs/reference/command-line-tools-reference/feature-gates/DRAExtendedResource)

Reviewer 3: ```suggestion
or having the [`DRAExtendedResource`](/docs/reference/command-line-tools-reference/feature-gates/#DRAExtendedResource)
```

feature gate enabled, disables the batching mechanism for those Pods to ensure correctness.

Note that you may need to review your `kube-scheduler` configuration
to ensure default topology spread constraints are not implicitly disabling batching for your workloads.

See the [docs](TODO) for more details about restrictions.

## The north star vision

> The scope of Workload Aware Scheduling is broad. These features are just the first steps.

Reviewer 3: ```suggestion
The project has a broad ambition to deliver workload aware scheduling. These new APIs and scheduling enhancements are just the first steps.
```

In the near future, the effort aims to tackle:

* Single-cycle workload scheduling
* Multi-node DRA and Topology Aware Scheduling
* Improved integration between scheduling and autoscaling
* Workload-level preemption
* Improved interaction with external workload schedulers
* Resource reservations
* Multi-workload scheduling simulations

And more. Stay tuned for further updates.

## Getting started

> You can enable the following features to try Workload Aware Scheduling:

Author: That's a good point, agree

* Workload API: Enable the `GenericWorkload` feature gate on both `kube-apiserver` and `kube-scheduler`,
  and ensure the `scheduling.k8s.io/v1alpha1`
  {{< glossary_tooltip text="API group" term_id="api-group" >}} is enabled.
* Gang Scheduling: Enable the `GangScheduling` feature gate on `kube-scheduler`
  (requires the Workload API to be enabled).

> * Opportunistic Batching: As a Beta feature, it is enabled by default in v1.35.

Reviewer 5: ```suggestion
* Opportunistic batching: As a Beta feature, it is enabled by default in v1.35.
```

>   You can disable it using the `OpportunisticBatching` feature gate on `kube-scheduler` if needed.

Reviewer 3: ```suggestion
  You can disable it using the [`OpportunisticBatching`](/docs/reference/command-line-tools-reference/feature-gates/#OpportunisticBatching) feature gate on `kube-scheduler` (if needed).
```

---
layout: blog
draft: true
title: "Kubernetes v1.35: Introducing Workload Aware Scheduling"
date: 2025-XX-XX
slug: introducing-workload-aware-scheduling
author: >
  Maciej Skoczeń (Google)
  Dominik Marciński (Google)
---

Scheduling large workloads is a much more complex and fragile operation than scheduling a single Pod,
as it often requires considering all Pods together instead of scheduling each one independently.
At the same time, the Pods that are part of such a workload are very often identical
from the scheduling perspective, which fundamentally changes how this process should look.

There are many custom schedulers adapted to perform workload scheduling efficiently,
but considering how common and important workload scheduling is to Kubernetes users,
it's high time to make workloads a first-class citizen for `kube-scheduler` and support them natively.

## Workload aware scheduling

> We are happy to announce the first elements of the *workload aware scheduling* effort that is aiming

Reviewer 3: ```suggestion
The recent 1.35 release of Kubernetes delivered the first tranche of *workload aware scheduling* improvements.
These are part of a wider effort that is aiming
```

Although _tranche_ isn't a widely known word, I am hoping that our readers can [glark](https://en.wiktionary.org/wiki/glark) the actual meaning from the rest of the sentence.

     the scheduler rejects **all** of the Pods in the group.
     They go back to the queue, freeing up the reserved resources for other workloads.

We'd like to point out that that while this is a first implementation, the Kubernetes project firmly
intends to improve and expand the gang scheduling algorithm in future releases.
Benefits we hope to deliver include a single-cycle scheduling phase for a whole gang,
workload-level preemption, and more, moving towards the north star goal.

## Opportunistic batching

In addition to explicit gang scheduling, v1.35 introduces *opportunistic batching*.
This is a Beta feature that improves scheduling latency for identical Pods.

Unlike gang scheduling, this feature does not require the Workload API
or any explicit opt-in on the user's part. It works opportunistically within the scheduler
by identifying Pods that have identical scheduling requirements (container images, resource requests,
affinities, etc.). When the scheduler processes a Pod, it can reuse the feasibility calculations
for subsequent identical Pods in the queue, significantly speeding up the process.

> 

Author: Added

### Restrictions

Opportunistic batching works under specific conditions. All fields used by the `kube-scheduler`
to find a placement must be identical between Pods. Additionally, using
[topology spread constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/),
[DRA ResourceClaims](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/)
or having the [`DRAExtendedResource`](/docs/reference/command-line-tools-reference/feature-gates/#DRAExtendedResource)
feature gate enabled, disables the batching mechanism for those Pods to ensure correctness.

Note that you may need to review your `kube-scheduler` configuration
to ensure default topology spread constraints are not implicitly disabling batching for your workloads.

See the [docs](TODO) for more details about restrictions.

## The north star vision

The project has a broad ambition to deliver workload aware scheduling.
These new APIs and scheduling enhancements are just the first steps.
In the near future, the effort aims to tackle:

* Single-cycle workload scheduling

> * Multi-node [DRA]((/docs/concepts/scheduling-eviction/dynamic-resource-allocation/))

Reviewer 3: Syntax looks wrong here

>   and Topology Aware Scheduling

Reviewer 3: nit: sentence case please (can hyperlink to the docs about topology aware scheduling if there are any)

* Improved integration between scheduling and autoscaling
* Workload-level preemption
* Improved interaction with external workload schedulers
* Resource reservations
* Multi-workload scheduling simulations

And more. Stay tuned for further updates.

## Getting started

> You can enable Kubernetes feature gates to try workload aware scheduling:

Author: You're right. Rephrased that

* Workload API: Enable the
  [`GenericWorkload`](/docs/reference/command-line-tools-reference/feature-gates/#GenericWorkload)
  feature gate on both `kube-apiserver` and `kube-scheduler`, and ensure the `scheduling.k8s.io/v1alpha1`
  {{< glossary_tooltip text="API group" term_id="api-group" >}} is enabled.
* Gang scheduling: Enable the 
  [`GangScheduling`](/docs/reference/command-line-tools-reference/feature-gates/#GangScheduling)
  feature gate on `kube-scheduler` (requires the Workload API to be enabled).
* Opportunistic batching: As a Beta feature, it is enabled by default in v1.35.
  You can disable it using the
  [`OpportunisticBatching`](/docs/reference/command-line-tools-reference/feature-gates/#OpportunisticBatching)
  feature gate on `kube-scheduler` if needed.

We encourage you to try out workload aware scheduling in your test clusters

> and share your experiences to help shape the future of Kubernetes scheduling.

Author: Good point, added a list of places where the feedback can be sent

---
layout: blog
draft: true
title: "Kubernetes v1.35: Introducing Workload Aware Scheduling"
date: 2025-XX-XX
slug: introducing-workload-aware-scheduling
author: >
  Maciej Skoczeń (Google),
  Dominik Marciński (Google)
---

Scheduling large workloads is a much more complex and fragile operation than scheduling a single Pod,
as it often requires considering all Pods together instead of scheduling each one independently.
At the same time, the Pods that are part of such a workload are very often identical
from the scheduling perspective, which fundamentally changes how this process should look.

There are many custom schedulers adapted to perform workload scheduling efficiently,
but considering how common and important workload scheduling is to Kubernetes users,

> it's high time to make workloads a first-class citizen for `kube-scheduler` and support them natively.

Reviewer 6: Let's mention why we start it now, why "its high time"? (e.g., growth of ML workload usecase, many gang/WAS implementation in the community etc)

Opportunistic batching works under specific conditions. All fields used by the `kube-scheduler`
to find a placement must be identical between Pods. Additionally, using
[topology spread constraints](/docs/concepts/scheduling-eviction/topology-spread-constraints/),
[DRA ResourceClaims](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/)
or having the [`DRAExtendedResource`](/docs/reference/command-line-tools-reference/feature-gates/#DRAExtendedResource)
feature gate enabled, disables the batching mechanism for those Pods to ensure correctness.

Note that you may need to review your `kube-scheduler` configuration
to ensure default topology spread constraints are not implicitly disabling batching for your workloads.

See the [docs](TODO) for more details about restrictions.

## The north star vision

The project has a broad ambition to deliver workload aware scheduling.
These new APIs and scheduling enhancements are just the first steps.
In the near future, the effort aims to tackle:

* Introducing a workload scheduling phase
* Improved support for multi-node [DRA](/docs/concepts/scheduling-eviction/dynamic-resource-allocation/)
  and topology aware scheduling

> * Workload-level preemption

Author: The workload-level preemption will be a separate feature (KEP) with additional Workload API changes
